Backend/main.py

```diff
@@ -54,48 +51,174 @@ class RouteResponse(BaseModel):
+    Handles API errors and returns the parsed JSON response.
+    Raises HTTPException for ORS errors or network issues.
+    """
+    ORS_API_URL = "https://api.openrouteservice.org/v2/directions/driving-car/geojson"
 
     headers = {
         "Authorization": ORS_TOKEN,
         "Content-Type": "application/json",
-        "Accept": "application/json, application/geo+json, application/gpx+xml, img/png; charset=utf-8"
+        "Accept": "application/json, application/geo+json"
     }
 
-    payload = {
-        "coordinates": ors_coords,
-    }
+    logger.info("Sending request to ORS API. Payload keys: %s", list(payload.keys()))
+    if "options" in payload:
+        logger.info("Payload options: %s", payload["options"])
 
     try:
-        # Using the driving-car profile, adjust if needed (e.g., driving-hgv, cycling-road)
-        response = requests.post(f"{ORS_URL}/directions/driving-car/geojson", json=payload, headers=headers)
-        response.raise_for_status()  # Raise an exception for bad status codes (4xx or 5xx)
-    except requests.RequestException as e:
-        # This handles network errors, timeouts, etc.
-        raise HTTPException(status_code=503, detail=f"Failed to connect to OpenRouteService: {str(e)}")
-    except requests.HTTPError as e:
-        # This handles API errors returned by ORS
-        try:
-            detail = response.json()
-        except requests.exceptions.JSONDecodeError:
-            detail = response.text  # Fallback if the error response isn't JSON
-        raise HTTPException(status_code=response.status_code, detail=detail)
+        response = requests.post(
+            ORS_API_URL,
+            json=payload,
+            headers=headers,
+            timeout=ORS_TIMEOUT_SECONDS
+        )
+        response.raise_for_status()  # Raises HTTPError for 4xx/5xx responses
+    except requests.exceptions.Timeout:
+        logger.error("ORS API request timed out after %s seconds.", ORS_TIMEOUT_SECONDS)
+        raise HTTPException(status_code=504, detail="Routing service request timed out.")
+    except requests.exceptions.RequestException as e:
+        # Other request errors (e.g., network connection, DNS failure)
+        logger.error("ORS API request failed: %s\nSent Payload: %s", e, payload)
+        raise HTTPException(status_code=502, detail=f"Routing service connection error: {e}")
 
     try:
-        data = response.json()
+        return response.json()
+    except requests.exceptions.JSONDecodeError as e:
+        logger.error(
+            "Failed to decode ORS API JSON response: %s\nResponse body (first 500 chars): %s",
+            e, response.text[:500]
+        )
+        raise HTTPException(status_code=502, detail="Routing service returned an invalid JSON response.")
 
-        geometry_coords = data["features"][0]["geometry"]["coordinates"]
 
-        return {"route": geometry_coords}
+def _extract_route_from_ors_response(ors_data: Dict[str, Any]) -> List[List[float]]:
```

Comment:
All of these micro parsings should definitely belong to their respective classes